src/System.Memory/src/System/SpanHelpers.T.cs

```diff
@@ -45,60 +45,76 @@ public static int IndexOf<T>(ref T searchSpace, int searchSpaceLength, ref T val
             return -1;
         }
 
-        public static int IndexOf<T>(ref T searchSpace, T value, int length)
+        public static unsafe int IndexOf<T>(ref T searchSpace, T value, int length)
             where T : struct, IEquatable<T>
         {
             Debug.Assert(length >= 0);
 
-            int index = -1;
-            int remainingLength = length;
-            while (remainingLength >= 8)
+            IntPtr index = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
+            while (length >= 8)
             {
-                if (value.Equals(Unsafe.Add(ref searchSpace, ++index)))
-                    goto Found;
-                if (value.Equals(Unsafe.Add(ref searchSpace, ++index)))
-                    goto Found;
-                if (value.Equals(Unsafe.Add(ref searchSpace, ++index)))
-                    goto Found;
-                if (value.Equals(Unsafe.Add(ref searchSpace, ++index)))
-                    goto Found;
-                if (value.Equals(Unsafe.Add(ref searchSpace, ++index)))
-                    goto Found;
-                if (value.Equals(Unsafe.Add(ref searchSpace, ++index)))
-                    goto Found;
-                if (value.Equals(Unsafe.Add(ref searchSpace, ++index)))
-                    goto Found;
-                if (value.Equals(Unsafe.Add(ref searchSpace, ++index)))
+                if (value.Equals(Unsafe.Add(ref searchSpace, index)))
                     goto Found;
+                if (value.Equals(Unsafe.Add(ref searchSpace, index + 1)))
+                    goto Found1;
+                if (value.Equals(Unsafe.Add(ref searchSpace, index + 2)))
+                    goto Found2;
+                if (value.Equals(Unsafe.Add(ref searchSpace, index + 3)))
+                    goto Found3;
+                if (value.Equals(Unsafe.Add(ref searchSpace, index + 4)))
+                    goto Found4;
+                if (value.Equals(Unsafe.Add(ref searchSpace, index + 5)))
+                    goto Found5;
+                if (value.Equals(Unsafe.Add(ref searchSpace, index + 6)))
+                    goto Found6;
+                if (value.Equals(Unsafe.Add(ref searchSpace, index + 7)))
+                    goto Found7;
 
-                remainingLength -= 8;
+                length -= 8;
```

Comment:
`length -= x;` can be calculated early, immediately after the while/if tests? Then its value is ready earlier in the pipeline for the next test?